### ISSUE TITLE
[BUG] FMTC prevents web compilation

### ISSUE THREAD

**Issue by stx** [JaffaKetchup/flutter_map_tile_caching#96]
### What is the bug?

If we update to 7.x we can no longer compile the project.

```
: Error: The integer literal 8585861370448577604 can't be represented exactly in JavaScript.
metadata.g.dart:18
Try changing the literal to something that can be represented in Javascript. In Javascript 8585861370448577536 is the nearest value that can be represented exactly.
  id: -8585861370448577604,
       ^^^^^^^^^^^^^^^^^^^
: Error: The integer literal 8117814053675000476 can't be represented exactly in JavaScript.
recovery.g.dart:18
Try changing the literal to something that can be represented in Javascript. In Javascript 8117814053675000832 is the nearest value that can be represented exactly.
  id: -8117814053675000476,
       ^^^^^^^^^^^^^^^^^^^
: Error: The integer literal 1365152130637522244 can't be represented exactly in JavaScript.
store_descriptor.g.dart:18
Try changing the literal to something that can be represented in Javascript. In Javascript 1365152130637522176 is the nearest value that can be represented exactly.
  id: 1365152130637522244,
      ^^^^^^^^^^^^^^^^^^^
: Error: The integer literal 5030120948284417748 can't be represented exactly in JavaScript.
tile.g.dart:18
Try changing the literal to something that can be represented in Javascript. In Javascript 5030120948284418048 is the nearest value that can be represented exactly.
  id: -5030120948284417748,
       ^^^^^^^^^^^^^^^^^^^
: Error: The integer literal 5953778071269117195 can't be represented exactly in JavaScript.
tile.g.dart:43
Try changing the literal to something that can be represented in Javascript. In Javascript 5953778071269116928 is the nearest value that can be represented exactly.
      id: 5953778071269117195,
          ^^^^^^^^^^^^^^^^^^^

: Error: The integer literal 0xcbf29ce484222325 can't be represented exactly in JavaScript.
tools.dart:14
Try changing the literal to something that can be represented in Javascript. In Javascript 0xcbf29ce484222000 is the nearest value that can be represented exactly.
    int hash = 0xcbf29ce484222325;
               ^^^^^^^^^^^^^^^^^^
Failed to compile application.
```

### What is the expected behaviour?

Being able to compile.

### Platforms Affected

Android, iOS, Windows, MacOS, Linux

### Severity

Fatal: Causes the application to crash

### Frequency

Consistently: Always occurs at the same time and location

### Requirements

- [X] I agree to follow this project's [Code of Conduct](https://github.com/jaffaketchup/flutter_map_tile_caching/blob/main/CODE_OF_CONDUCT.md)
- [X] I have checked for known issues in the GitHub tracker and [on the docs site](https://fmtc.jaffaketchup.dev/known-issues-and-reporting#known-issues)
- [X] I have completed the pre-report checks listed [on the docs site](https://fmtc.jaffaketchup.dev/known-issues-and-reporting#reporting-a-bug)
- [X] If I am using FMTC in a proprietary project, I have read and understood [this important legal notice](https://fmtc.jaffaketchup.dev/#proprietary-licensing)

**Comment by JaffaKetchup** [comment]
Hi @stx,
This is a major oversight on my part. Of course FMTC isn't intended for use on web, but that doesn't mean it shouldn't compile.
The issue lies with the reliance on Isar, which uses a code generator to generate type safe database models.
The current workaround is to use conditional imports. But, as stated by the author of Isar in https://github.com/dart-lang/sdk/issues/50385, they are cumbersome to use. There is a proposal for conditionally imported libraries at https://github.com/dart-lang/language/issues/2235, but I doubt that will be merged any time soon.
For the time being, I'll have to look into conditional imports, but that might take a while (major refactoring). For now, I'm afraid you'll have to stick to v6, my apologies.
I'll try to have this resolved by the end of the week, but I am extremely busy at the moment. Thanks for your patience!

**Comment by JaffaKetchup** [comment]
@stx Unfortunatley, due to the way code generators work, it may not be possible to use conditional imports here:

![image](https://user-images.githubusercontent.com/58115698/219157952-b7c49e78-9f5e-4ffc-b4af-5b31125f6595.png)

Therefore, we're a little stuck for now. Apologies. If you think there's a way to solve this, please do let me know!

**Comment by JaffaKetchup** [comment]
See https://github.com/isar/isar/issues/686.

**Comment by JaffaKetchup** [comment]
Just heard a little bit from Isar's maintainer: https://github.com/isar/isar/issues/686#issuecomment-1493325747. This is great news!
v8 will support Isar v4 and flutter_map v4, and will not be released before those.

**Comment by JaffaKetchup** [comment]
See https://github.com/isar/isar/issues/686.

**Comment by wikylyu** [comment]
I have the same issue.

**Comment by tik1111** [comment]
any progress yet? my web app wont compile now :S

**Comment by JaffaKetchup** [comment]
Hey,
Isar v4 is under active development and appears to be almost ready. I'm concentrating on the next release of FM for the moment, but then this should be coming soon!

**Comment by slawekkrol** [comment]
There is a temporary solution for that:

1. Create 3 files:
<img width="358" alt="Zrzut ekranu 2023-09-30 o 12 47 49" src="https://github.com/JaffaKetchup/flutter_map_tile_caching/assets/5537577/e97a2c20-9e37-488b-b555-f3cf93b98848">

2. Add to map_cache and map_cache_web:

```
import 'package:flutter_map/flutter_map.dart';

Future<void> initializeCache() async {}

TileProvider? tileProvider;
```

3. Add to map_cache_mobile:

```
import 'package:flutter_map/flutter_map.dart';
import 'package:flutter_map_tile_caching/flutter_map_tile_caching.dart';

const cacheStore = 'mapStore';

Future<void> initializeCache() async {
  await FlutterMapTileCaching.initialise();
  await FMTC.instance(cacheStore).manage.createAsync();
}

TileProvider? tileProvider = FMTC.instance(cacheStore).getTileProvider();
```

4. When you using map cache add conditional import:

```
import 'your_path_to/cache/map_cache.dart'
    if (dart.library.io) 'your_path_to/cache/map_cache_mobile.dart'
    if (dart.library.html) 'your_path_to/cache/map_cache_web.dart';
```

And use functions according to doc initializeCache on app start and tileProvider on map

**Comment by Soap-141** [comment]
> There is a temporary solution for that:
> 
> 1. Create 3 files:
> 
> <img alt="Zrzut ekranu 2023-09-30 o 12 47 49" width="358" src="https://private-user-images.githubusercontent.com/5537577/271768937-e97a2c20-9e37-488b-b555-f3cf93b98848.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwNDEzMDYsIm5iZiI6MTcyMzA0MTAwNiwicGF0aCI6Ii81NTM3NTc3LzI3MTc2ODkzNy1lOTdhMmMyMC05ZTM3LTQ4OGItYjU1NS1mM2NmOTNiOTg4NDgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDdUMTQzMDA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDkwMzUyNTY3NTQyMDY4NzRjODU2MDk3ZWM3NTJhMTQ0OTViOTAwOGMyMGE1YzkzOGU2OTUwNGRhZWQ3MWNkNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.v9HUhgd9-PXkN-OZPBO9RdBOokqtSs_PIL-U9BuvdtA">
> 2. Add to map_cache and map_cache_web:
> 
> ```
> import 'package:flutter_map/flutter_map.dart';
> 
> Future<void> initializeCache() async {}
> 
> TileProvider? tileProvider;
> ```
> 
> 3. Add to map_cache_mobile:
> 
> ```
> import 'package:flutter_map/flutter_map.dart';
> import 'package:flutter_map_tile_caching/flutter_map_tile_caching.dart';
> 
> const cacheStore = 'mapStore';
> 
> Future<void> initializeCache() async {
>   await FlutterMapTileCaching.initialise();
>   await FMTC.instance(cacheStore).manage.createAsync();
> }
> 
> TileProvider? tileProvider = FMTC.instance(cacheStore).getTileProvider();
> ```
> 
> 4. When you using map cache add conditional import:
> 
> ```
> import 'your_path_to/cache/map_cache.dart'
>     if (dart.library.io) 'your_path_to/cache/map_cache_mobile.dart'
>     if (dart.library.html) 'your_path_to/cache/map_cache_web.dart';
> ```
> 
> And use functions according to doc initializeCache on app start and tileProvider on map
> 
> ```
> import 'your_path_to/cache/map_cache.dart'
>     if (dart.library.io) 'your_path_to/cache/map_cache_mobile.dart'
>     if (dart.library.html) 'your_path_to/cache/map_cache_web.dart';
> 
> void main() async {
>   WidgetsFlutterBinding.ensureInitialized();
>   await initializeCache();
>   runApp(const App());
> }
> ```
> 
> ```
> import 'your_path_to/cache/map_cache.dart'
>     if (dart.library.io) 'your_path_to/cache/map_cache_mobile.dart'
>     if (dart.library.html) 'your_path_to/cache/map_cache_web.dart';
> 
> TileLayer(
>                         urlTemplate: my_url_template,
>                         userAgentPackageName: 'my_app',
>                         tileProvider: tileProvider,
>                       )
> ```

This doesn't fix the issue, the package is still used in the project and it doesn't compile. @slawekkrol

**Comment by JaffaKetchup** [comment]
@Soap-141 What's the compile error with that solution?

**Comment by Soap-141** [comment]
@JaffaKetchup It's the same thing, I haven't tried with the latest package version yet because of the iOS issue, my team is not ready to update until it's fixed, and now the client wants a Web version of the application. I'm doing a 'Spike' to figure out what will need to change in the application to make it work on the Web.

**Comment by Soap-141** [comment]
One of them might be on us, but I've got multiple from the package itself.

**Comment by JaffaKetchup** [comment]
Can you post the exact compilation error please? Also note that even if it compiles, FMTC won't work.

**Comment by Soap-141** [comment]
> Can you post the exact compilation error please? Also note that even if it compiles, FMTC won't work.

Sure, give me 15 minutes, and yes I know, no offlines map for the Web anyway, we'll use the other caching package.

**Comment by Soap-141** [comment]
Launching lib\main.dart on Chrome in debug mode...
Warning: In index.html:35: Local variable for "serviceWorkerVersion" is deprecated. Use "{{flutter_service_worker_version}}" template token instead.
Warning: In index.html:44: "FlutterLoader.loadEntrypoint" is deprecated. Use "FlutterLoader.load" instead.
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/db/defs/metadata.g.dart:18:8: Error: The integer literal 8585861370448577604 can't be represented exactly in JavaScript.
Try changing the literal to something that can be represented in JavaScript. In JavaScript 8585861370448577536 is the nearest value that can be represented exactly.
  id: -8585861370448577604,
       ^^^^^^^^^^^^^^^^^^^
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/db/defs/recovery.g.dart:18:8: Error: The integer literal 8117814053675000476 can't be represented exactly in JavaScript.
Try changing the literal to something that can be represented in JavaScript. In JavaScript 8117814053675000832 is the nearest value that can be represented exactly.
  id: -8117814053675000476,
       ^^^^^^^^^^^^^^^^^^^
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/db/defs/store_descriptor.g.dart:18:7: Error: The integer literal 1365152130637522244 can't be represented exactly in JavaScript.
Try changing the literal to something that can be represented in JavaScript. In JavaScript 1365152130637522176 is the nearest value that can be represented exactly.
  id: 1365152130637522244,
      ^^^^^^^^^^^^^^^^^^^
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/db/defs/tile.g.dart:18:8: Error: The integer literal 5030120948284417748 can't be represented exactly in JavaScript.
Try changing the literal to something that can be represented in JavaScript. In JavaScript 5030120948284418048 is the nearest value that can be represented exactly.
  id: -5030120948284417748,
       ^^^^^^^^^^^^^^^^^^^
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/db/defs/tile.g.dart:43:11: Error: The integer literal 5953778071269117195 can't be represented exactly in JavaScript.
Try changing the literal to something that can be represented in JavaScript. In JavaScript 5953778071269116928 is the nearest value that can be represented exactly.
      id: 5953778071269117195,
          ^^^^^^^^^^^^^^^^^^^
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/misc/int_extremes.dart:4:20: Error: The integer literal 9223372036854775807 can't be represented exactly in JavaScript.
Try changing the literal to something that can be represented in JavaScript. In JavaScript 9223372036854775808 is the nearest value that can be represented exactly.
const largestInt = 9223372036854775807;
                   ^^^^^^^^^^^^^^^^^^^
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/bulk_download/download_progress.dart:154:20: Error: The argument type 'num' can't be assigned to the parameter type 'int'.
                  .clamp(elapsedDuration.inSeconds, largestInt),
                   ^
/C:/Users/thomas.labrecque/AppData/Local/Pub/Cache/hosted/pub.dev/flutter_map_tile_caching-9.0.0-dev.5/lib/src/db/tools.dart:16:16: Error: The integer literal 0xcbf29ce484222325 can't be represented exactly in JavaScript.
Try changing the literal to something that can be represented in JavaScript. In JavaScript 0xcbf29ce484222000 is the nearest value that can be represented exactly.
    int hash = 0xcbf29ce484222325;
               ^^^^^^^^^^^^^^^^^^
Failed to compile application.


Exited (1).

**Comment by JaffaKetchup** [comment]
Fixed in v9.1.2 by #165. Fix will also be available in v10 from next prerelease (after >= v10.0.0-dev.4).
